### PR TITLE
[FLINK-15334][table sql / api] Fix physical schema mapping in TableFormatFactoryBase to support define orderless computed column

### PR DESCRIPTION
**Pull request by leonardBang**
Fix physical schema mapping in TableFormatFactoryBase to support define orderless computed column

## What is the purpose of the change

*This pull request Fix physical schema mapping in schemaValidator which used in TableSoure and tableSourceFactory, we can support  define orderless computed column after fix it.*

## Brief change log

  - *fix mapping in org/apache/flink/table/factories/TableFormatFactoryBase.java*
  - *Add unit test in  file org/apache/flink/table/descriptors/SchemaValidatorTest.scala*
  - *Add IT case org/apache/flink/streaming/connectors/kafka/KafkaTableTestBase.java*

## Verifying this change

  - *Add unit test cover path that define table by DescriptorProperties*
  - *Add IT case cover path that define table by DDL*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): ( no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/factories/TableFormatFactoryBase.java [hunk 1]
```diff
@@ -154,31 +156,37 @@ public static TableSchema deriveSchema(Map<String, String> properties) {
...
 
-		final TableSchema baseSchema = TableSchemaUtils.getPhysicalSchema(
-			descriptorProperties.getTableSchema(SCHEMA));
-		for (int i = 0; i < baseSchema.getFieldCount(); i++) {
-			final String fieldName = baseSchema.getFieldNames()[i];
-			final DataType fieldType = baseSchema.getFieldDataTypes()[i];
+		final TableSchema tableSchema = descriptorProperties.getTableSchema(SCHEMA);
+		final TableSchema physicalSchema = TableSchemaUtils.getPhysicalSchema(tableSchema);
+
+		final Map<Integer, Integer> physicalIndices2Indices = Arrays.stream(physicalSchema.getFieldNames())
```
Comment (wuchong) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/factories/TableFormatFactoryBase.java]: A simper way to fix is add the field only when it is not generated. We already have `isGeneratedColumn `. So we don't need the complex mapping.

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/descriptors/SchemaValidatorTest.scala [hunk 2]
```diff
@@ -168,25 +168,28 @@ class SchemaValidatorTest {
...
+    properties.putString("schema.0.data-type", DataTypes.STRING().toString)
+    properties.putString("schema.1.name", "computed-column")
+    properties.putString("schema.1.data-type", DataTypes.INT().toString)
+    properties.putString("schema.1.expr", "f4 + 1")
+    properties.putString("schema.2.name", "f2")
+    properties.putString("schema.2.data-type", DataTypes.INT().toString)
+    properties.putString("schema.3.name", "f3")
+    properties.putString("schema.3.data-type", DataTypes.TIMESTAMP(3).toString)
+    properties.putString("schema.4.name", "f4")
+    properties.putString("schema.4.data-type", DataTypes.INT().toString)
```
Comment (wuchong) [inline flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/descriptors/SchemaValidatorTest.scala]: Please also add `Rowtime` properties in it to verify the logic of `SchemaValidator#deriveTableSinkSchema`. 
I think `SchemaValidator#deriveTableSinkSchema` still has bug.

File: flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/descriptors/SchemaValidator.java [hunk 3]
```diff
@@ -212,12 +212,16 @@ else if (proctimeFound) {
...
-
-		for (int i = 0; i < schema.getFieldCount(); i++) {
-			TypeInformation t = schema.getFieldTypes()[i];
-			String n = schema.getFieldNames()[i];
+		TableSchema tableSchema = properties.getTableSchema(SCHEMA);
+		for (int i = 0; i < tableSchema.getFieldCount(); i++) {
+			TypeInformation t = tableSchema.getFieldTypes()[i];
+			String n = tableSchema.getFieldNames()[i];
+			Optional<TableColumn> tableColumn = tableSchema.getTableColumn(n);
+			boolean isGeneratedColumn = tableColumn.isPresent() && tableColumn.get().isGenerated();
```
Comment (wuchong) [inline flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/descriptors/SchemaValidator.java]: Improve the code a bit:

```java
final TableColumn tableColumn = tableSchema.getTableColumns().get(i);
final String fieldName = tableColumn.getName();
final DataType fieldType = tableColumn.getType();
final boolean isGeneratedColumn = tableColumn.isGenerated();
```
We do know it's safe to call `tableSchema.getTableColumns().get(i)`.

Comment (leonardBang) [reply flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/descriptors/SchemaValidator.java]: That's better, thanks for your kindly tips

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/factories/TableFormatFactoryBase.java [hunk 4]
```diff
@@ -157,36 +156,33 @@ public static TableSchema deriveSchema(Map<String, String> properties) {
...
-		for (int i = 0; i < physicalSchema.getFieldCount(); i++) {
-			final String fieldName = physicalSchema.getFieldNames()[i];
-			final DataType fieldType = physicalSchema.getFieldDataTypes()[i];
-
+		for (int i = 0; i < tableSchema.getFieldCount(); i++) {
+			final String fieldName = tableSchema.getFieldNames()[i];
+			final DataType fieldType = tableSchema.getFieldDataTypes()[i];
+
+			final Optional<TableColumn> tableColumn = tableSchema.getTableColumn(fieldName);
+			final boolean isGeneratedColumn = tableColumn.isPresent() && tableColumn.get().isGenerated();
```
Comment (wuchong) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/factories/TableFormatFactoryBase.java]: Please update here as above.

Comment (leonardBang) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/factories/TableFormatFactoryBase.java]: fixed

**Comment by leonardBang** [comment]
@wuchong thanks!

**Pull request merged**